stan/math/rev/fun/diag_pre_multiply.hpp

```diff
@@ -0,0 +1,92 @@
+#ifndef STAN_MATH_REV_FUN_DIAG_PRE_MULTIPLY_HPP
+#define STAN_MATH_REV_FUN_DIAG_PRE_MULTIPLY_HPP
+
+#include <stan/math/prim/fun/Eigen.hpp>
+#include <stan/math/prim/err.hpp>
+#include <stan/math/rev/core.hpp>
+
+namespace stan {
+namespace math {
+
+/**
+ * Return the product of the diagonal matrix formed from the vector
+ * or row_vector and a matrix.
+ *
+ * @tparam T1 type of the vector/row_vector
+ * @tparam T2 type of the matrix
+ * @param m1 input vector/row_vector
+ * @param m2 input matrix
+ *
+ * @return product of the diagonal matrix formed from the
+ * vector or row_vector and a matrix.
+ */
+template <typename T1, typename T2, require_eigen_vector_t<T1>* = nullptr,
+          require_eigen_t<T2>* = nullptr,
+          require_any_rev_matrix_t<T1, T2>* = nullptr>
```

Comment:
I think if you change these requires to:

```cpp
template <typename T1, typename T2,
          require_vector_t<T1>* = nullptr,
          require_matrix_t<T2>* = nullptr,
          require_any_st_var<T1, T2>* = nullptr>
```

Then you're good for varmat too if I'm reading the code right. `require_vector_t` actually allows `std::vector` as well but we just won't expose that signature.

To test varmat, for every `stan::test::expect_ad` add a `stan::test::expect_ad_matvar` with the same arguments.

Reply:
Great, thank you!

Reply:
I implemented this and it works OK when I try it manually: Create a main, define a couple of variables, lp, and compute the gradient. For example this works OK:

```
int main() {
	Eigen::Matrix<stan::math::var, 1, 2> M1;
	Eigen::Matrix<stan::math::var, 2, 2> M2;
	Eigen::Matrix<double, 2, 2> M2;
	stan::math::var_value<Eigen::Matrix<double, 2, 2>> M_res1;
	M1 << 1,2;
	M2 << 2,3,2,5;
	M_res1 = diag_pre_multiply(M1, M2);
	stan::math::var lp2 = 0;
	for (int i = 0; i < 2; ++i) {
		for (int j = 0; j < 2; ++j) {
			lp2 += M_res1(i,j);
		}
	}
	lp2.grad();
	std::cout << M1.adj() << std::endl;
	std::cout << M2.adj() << std::endl;
	return 0;
}
```
However, when I try running the mix tests I still get a bunch of errors. I think the two most relevant are:

```
./test/unit/math/test_ad_matvar.hpp:467:23: error: no matching function for call to 'test_matvar_gradient(const stan::test::ad_tolerances&, T_mv_ret&, T_vm_ret&, std::tuple<Eigen::Matrix<double, -1, 1, 0, -1, 1>, Eigen::Matrix<stan::math::var_value<double, void>, -1, -1, 0, -1, -1> >&, std::tuple<Eigen::Matrix<double, -1, 1, 0, -1, 1>, stan::math::var_value<Eigen::Matrix<double, -1, -1, 0, -1, -1>, void> >&)'
   test_matvar_gradient(tols, A_mv_ret, A_vm_ret, A_mv_tuple, A_vm_tuple);
```

and

`./stan/math/rev/functor/gradient.hpp:47:17: error: conversion from 'var_value<Eigen::Matrix<double, -1, -1>>' to non-scalar type 'var_value<double>' requested`

I am not that familiar with these tests. Is there a bug in my implementation, or is something else missing? I looked at a couple of other rev implementations that use `make_callback_var` and I couldn't find any obvious differences to my implementation.

Reply:
If you're getting this error with reverse_psas_callback or what I said doesn't make sense, push your code and I'll have a look.

Reply:
So that means I should use the initial implementation, right?

Reply:
Yeah for this code:

```
Eigen::Matrix<stan::math::var, 1, 2> M1;
Eigen::Matrix<stan::math::var, 2, 2> M2;
Eigen::Matrix<double, 2, 2> M2;
stan::math::var_value<Eigen::Matrix<double, 2, 2>> M_res1;
...
M_res1 = diag_pre_multiply(M1, M2);
```

we're assuming for stanc3 that if all the input types are matvar, then the output will be matvar as well. So the fact that this returns a varmat is violatin' assumptions.

Reply:
Yeah use the `reverse_pass_callback` for the contents, but use these new requires to enable varmat support.

Reply:
> If you're getting this error with reverse_psas_callback or what I said doesn't make sense, push your code and I'll have a look.

No, with `reverse_pass_callback` it works (all tests, including OpenCL). I was just trying to do it with `make_callback_var`, as you proposed. OK, I'll revert to that implementation and push, thanks!